src/installer/iohandler.py

```diff
@@ -0,0 +1,100 @@
+"""Handling wheel installation I/O-related functionality."""
+
+import hashlib
+import os
+from io import RawIOBase
+
+from installer._compat.typing import TYPE_CHECKING
+
+from .records import Hash, Record
+
+if TYPE_CHECKING:
+    from ._compat.typing import FSPath
+    from typing import Dict, Tuple, Optional, Callable
+
+_BLOCK_SIZE = 4096
+_SCHEME_KEYS = ["platlib", "purelib"]
+
+
+def basic_copy_handler(
+    source,  # type: RawIOBase
+    dest,  # type: FSPath
+    hash_algorithm="sha256",  # type: Optional[str]
+):
+    # type: (...) -> Tuple[int, Hash]
+    """Copy a buffer and compute its hash and size.
+
+    Copies the source buffer to the specified destination. It also computes the hash
+    of the file to avoid traversing the source twice.
+
+    :param source: buffer holding the source data
+    :param dest: destination path
+    :param hash_algorithm: hashing algorithm
+
+    :return: size and ``Hash`` tuple of the copied buffer
+    """
+    assert hash_algorithm is not None  # to please mypy
+    hasher = hashlib.new(hash_algorithm)
+    size = 0
+    with open(dest, "wb") as dest_fd:
```

Comment:
I'd prefer we open the file in the `IOHandler`, so that this function isn't tied to the filesystem.

Reply:
But then IOHandler will be tied to the FS. Although `os.makedirs` ties it to it too, but I want to move that into the copy handler as well.

Reply:
I'd rather have that be tied to the file system in a specific method within it, rather than this function by tied to the file system -- this can directly consume streams, which is useful when you're transforming the contents of the wheel (eg: shebang rewriting).